src/WebJobs.Script/IO/AutoRecoveringFileSystemWatcher.cs

```diff
@@ -0,0 +1,202 @@
+        {
+            _path = path;
+            _filter = filter;
+            _changeTypes = changeTypes;
+            _includeSubdirectories = includeSubdirectories;
+            _traceWriter = traceWriter;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _cancellationToken = _cancellationTokenSource.Token;
+            _handleFileError = new Action<ErrorEventArgs>(OnFileWatcherError).Debounce();
+            
+            InitializeWatcher();
+        }
+      
+        ~AutoRecoveringFileSystemWatcher()
+        {
+            Dispose(false);
+        }
+
+        private void InitializeWatcher()
+        {
+            if (!Directory.Exists(_path))
+            {
+                throw new DirectoryNotFoundException($"The path '{_path}' cannot be found.");
+            }
+
+            lock (_syncRoot)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                _fileWatcher = new FileSystemWatcher(_path, _filter)
+                {
+                    IncludeSubdirectories = _includeSubdirectories,
+                    EnableRaisingEvents = true
+                };
+
+                _fileWatcher.Error += (s, a) => _handleFileError(a);
+                AddEventSubscriptions(_fileWatcher, _changeTypes);
+            }
+        }
+
+        private void AddEventSubscriptions(FileSystemWatcher fileWatcher, WatcherChangeTypes changeTypes)
+        {
+            if (changeTypes.HasFlag(WatcherChangeTypes.Changed))
+            {
+                fileWatcher.Changed += OnFileChanged;
+            }
+
+            if (changeTypes.HasFlag(WatcherChangeTypes.Created))
+            {
+                fileWatcher.Created += OnFileChanged;
+            }
+
+            if (changeTypes.HasFlag(WatcherChangeTypes.Deleted))
+            {
+                fileWatcher.Deleted += OnFileChanged;
+            }
+
+            if (changeTypes.HasFlag(WatcherChangeTypes.Renamed))
+            {
+                fileWatcher.Renamed += OnFileChanged;
+            }
+        }
+
+        private void RemoveEventSubscriptions(FileSystemWatcher fileWatcher)
+        {
+            fileWatcher.Changed -= OnFileChanged;
+            fileWatcher.Created -= OnFileChanged;
+            fileWatcher.Deleted -= OnFileChanged;
+            fileWatcher.Renamed -= OnFileChanged;
+        }
+
+        protected void OnFileChanged(object sender, FileSystemEventArgs e)
+        {
+            Changed?.Invoke(this, e);
+        }
+
+        protected void OnFileWatcherError(ErrorEventArgs obj)
+        {
+            Trace($"Failure detected '{obj.GetExceptionMessage()}'. Initiating recovery...", TraceLevel.Warning);
+
+            Recover().ContinueWith(t =>
+            {
+                if (t.IsFaulted || t.IsCanceled)
+                {
+                    t.Exception?.Handle(e => true);
+                    Trace($"Recovery process aborted.", TraceLevel.Error);
```

Comment:
With this implementation, this should only happen if this type is disposed/cancellation is requested. So, in theory, it should only happen when there would be no need for events anymore.

If if does happen, though, we do give up, which would put us in the state we end up in today, the key difference is that we'd at least be able to detect that scenario by analyzing logs.